### DISCUSSION TITLE
An exception was thrown during the encryption key exchange.

### DISCUSSION THREAD

**Discussion by policyscientist** [sassoftware/saspy#505] [Q&A] [answered]
I am using SAS ODA with saspy in Google COLAB environment. After the recent update of ODA, establishing a connection with SAS is failed due to an authentication error. Since the log-in procedure has no problem on the SAS ODA webpage, it might be troubled with saspy. Please help me solve this problem. Thanks.

code :

import saspy
sas=saspy.SASsession(java='/usr/bin/java', iomhost=['odaws01-apse1.oda.sas.com','odaws02-apse1.oda.sas.com','odaws01-apse1-2.oda.sas.com','odaws02-apse1-2.oda.sas.com'], iomport=8591, omruser='policyscientist@gmail.com', omrpw='**deleted**', encoding='utf-8')

output :

Using SAS Config named: default
We failed in getConnection
An exception was thrown during the encryption key exchange.

SAS process has terminated unexpectedly. Pid State= (2817, 64000)
SAS Connection failed. No connection established. Double check your settings in sascfg_personal.py file.

Attempted to run program /usr/bin/java with the following parameters:['/usr/bin/java', '-classpath', '/usr/local/lib/python3.8/dist-packages/saspy/java/saspyiom.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/log4j-1.2-api-2.12.4.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/log4j-api-2.12.4.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/log4j-core-2.12.4.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/sas.security.sspi.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/sas.core.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/sas.svc.connection.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/sas.rutil.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/sas.rutil.nls.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/sastpj.rutil.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/thirdparty/glassfish-corba-internal-api.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/thirdparty/glassfish-corba-omgapi.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/thirdparty/glassfish-corba-orb.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/thirdparty/pfl-basic.jar:/usr/local/lib/python3.8/dist-packages/saspy/java/thirdparty/pfl-tf.jar', 'pyiom.saspy2j', '-host', 'localhost', '-stdinport', '44355', '-stdoutport', '39215', '-stderrport', '38273', '-iomhost', 'odaws01-apse1.oda.sas.com;odaws02-apse1.oda.sas.com;odaws01-apse1-2.oda.sas.com;odaws02-apse1-2.oda.sas.com', '-iomport', '8591', '-user', '[policyscientist@gmail.com](mailto:policyscientist@gmail.com)', '-lrecl', '1048576', '']

No SAS process attached. SAS process has terminated unexpectedly.
Invalid response from SAS on inital submission. printing the SASLOG as diagnostic

---------------------------------------------------------------------------
SASIOConnectionError                      Traceback (most recent call last)
[<ipython-input-4-0ac19ff15a87>](https://localhost:8080/#) in <module>
      1 import saspy
----> 2 sas=saspy.SASsession(java='/usr/bin/java', iomhost=['odaws01-apse1.oda.sas.com','odaws02-apse1.oda.sas.com','odaws01-apse1-2.oda.sas.com','odaws02-apse1-2.oda.sas.com'], iomport=8591, omruser='[policyscientist@gmail.com](mailto:policyscientist@gmail.com)', omrpw='**deleted**', encoding='utf-8')

[/usr/local/lib/python3.8/dist-packages/saspy/sasbase.py](https://localhost:8080/#) in __init__(self, **kwargs)
    606 
    607            if failed:
--> 608               raise SASIOConnectionError(res)
    609 
    610            if self.pyenc is not None:

SASIOConnectionError: Failure establishing SASsession.
 No SAS process attached. SAS process has terminated unexpectedly.

**Comment by tomweber-sas** [comment, marked as answer]
Hey, given the error, it would seem you need the encryption jars now, for this to be able to connect. I can't connect to the AP region to verify, but trying to connect to the US region I can get at, it seems they've upgraded to M7, which requires the encryption jars to be in the classpath. Take a look at the doc here on this: https://sassoftware.github.io/saspy/configuration.html#attn-as-of-saspy-version-3-3-3-the-classpath-is-no-longer-required-in-your-configuration-file

You will need to add the 3 jar file to the path where saspy is installed; into this directory (based upon your output above): /usr/local/lib/python3.8/dist-packages/saspy/java/iomclient/

Assuming you don't have access to a local SAS deployment, where you can get these 3 jars from, you can download them from this link: https://support.sas.com/downloads/package.htm?pid=2494

I expect that once you add these, you will be able to connect. Since it looks like they have upgraded to M7, I'll see about getting the documentation updated to identify that this is now a requirement, when it previously wasn't. SAS is closed until the new year, so this will have to wait until we're back at work. But for now, you should be able to get around this by just adding the jars. 

One other thought, if you can't write the jar files to that location (don't have permission or something), you can copy them to wherever you can, and then you would have to provide the classpath in your config; just copy the whole classpath from the output you have above, and change the paths for those 3 jars to wherever you were able to write them; keeping all of the other paths as is. Hopefully you can just copy the jars to the right location to begin with and it should all work.

Thanks,
Tom

**Reply by policyscientist** [reply]
I can write the jar files to the location you mentioned in the COLAB environment without permission issues. Thanks for your help.

**Reply by tomweber-sas** [reply]
That's great, I'm glad that was it! I do see other info saying they upgraded ODA to use M7, so I'll have to get that doc updated with this info, as it will hit a lot of other people!
Thanks,
Tom

**Reply by mrmorgan17** [reply]
Thank you! This solved my issue as well
